oml/functional/metrics.py

```diff
@@ -220,42 +216,209 @@ def calculate_accuracy_on_triplets(embeddings: torch.Tensor, reduce_mean: bool =
+
+    ``cmc@k`` is 1 if ``top_k`` distances contain an element from the same class, and 0 otherwise.
+
+    Args:
+        gt_tops: ground truth of the ``max(top_k)`` distances.
+        top_k: Tuple of ``k`` values to calculate ``cmc@k`` for.
+
+    Returns:
+        List of ``cmc@k`` tensors.
+
+    Example:
+        >>> gt_tops = torch.tensor([[1, 0, 0], [0, 1, 1], [0, 0, 1]], dtype=torch.bool)
+        >>> calc_cmc(gt_tops, (1, 2, 3))
+        [tensor([1., 0., 0.]), tensor([1., 1., 0.]), tensor([1., 1., 1.])]
+    """
+    _check_if_nonempty_integers_and_positive(top_k, "top_k")
+    top_k = _clip_max_with_warning(top_k, gt_tops.shape[1])
+    cmc = []
+    for k in top_k:
+        cmc.append(torch.any(gt_tops[:, :k], dim=1).float())
+    return cmc
+
+
+def calc_precision(gt_tops: torch.Tensor, n_gt: torch.Tensor, top_k: Tuple[int, ...]) -> List[torch.Tensor]:
+    """
+    Function to compute Precision@k for each sample.
+
+    ``precision@k`` is a proportion of the ``top_k`` elements from the same class to total number of the elements
+    from that class.
+
+    Args:
+        gt_tops: ground truth of the at ``max(top_k)`` distances.
+        n_gt: total number of samples from the same class as a query sample.
+        top_k: Tuple of ``k`` values to calculate ``cmc@k`` for.
+
+    Returns:
+        List of ``precision@k`` tensors.
+
+    Given a list of ground truth top :math:`k` elements from the gallery :math:`[g_1, \\ldots, g_k]` and
+    total number of elements from the same class in the gallery :math:`n`
+    the :math:`\\textrm{Precision}@k` for the element is defined as
+
+    .. math::
+        \\textrm{Precition}@k = \\frac{1}{\\min{\\left(k, n\\right)}}\\sum\\limits_{i = 1}^k g_i
+
+    See:
+        `Evaluation measures (information retrieval). Precision`_
+
+
+        .. _`Evaluation measures (information retrieval). Precision`:
+            https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Precision
+
+    Example:
+        >>> gt_tops = torch.tensor([[1, 0, 0], [0, 1, 1], [0, 0, 1]], dtype=torch.bool)
+        >>> n_gt = torch.tensor([2, 100, 5])
+        >>> calc_precision(gt_tops, n_gt, (1, 2, 3))
+        [tensor([1., 0., 0.]), tensor([0.5000, 0.5000, 0.0000]), tensor([0.5000, 0.6667, 0.3333])]
+    """
+    _check_if_nonempty_integers_and_positive(top_k, "top_k")
+    top_k = _clip_max_with_warning(top_k, gt_tops.shape[1])
+    precision = []
+    correct_preds = torch.cumsum(gt_tops.float(), dim=1)
+    for k in top_k:
+        _n_gt = torch.min(n_gt, torch.tensor(k).unsqueeze(0))
+        precision.append(correct_preds[:, k - 1] / _n_gt)
+    return precision
+
+
+def calc_map(gt_tops: torch.Tensor, n_gt: torch.Tensor, top_k: Tuple[int, ...]) -> List[torch.Tensor]:
+    """
+    Function to compute Mean Average Precision (MAP) for each sample.
+
+    ``map@k`` is the mean value of the ``precision@1``, ``precision@2``, ..., ``precision@k``
```

Comment:
Could you give me a link on some article on MAP that you describe? All the articles I saw describe $map@k$ as

$map@k = \frac{1}{n}\sum\limits_{i = 1}^{k} precision@i \times rel(i)$,

where $rel(i)$ is 1 if $i$-th sample from the gallery relevant to the query an 0 otherwise.

See 
https://chaitanyabelhekar.medium.com/recommender-system-metrics-clearly-explained-1f2ba6690216
https://www.kaggle.com/code/nandeshwar/mean-average-precision-map-k-metric-explained-code/notebook
https://neptune.ai/blog/recommender-systems-metrics
https://machinelearninginterview.com/topics/machine-learning/mapatk_evaluation_metric_for_ranking/
https://www.coursera.org/lecture/basic-recommender-systems/ranking-metrics-F1gLc

Following your example, the $map@6$ calculated by the formula from links above shoud be

```
map@6 = 1/3 *(1/1 + 2/3 + 3/3)
```

Reply:
Misunderstanding appears in the definition of `precision@k` in our repo and other resources. 

Usually, people define precision as `precision(k) = cumsum(gt)(k) / k` and `recall(k) = cumsum(gt)(k) / n_gt`.

But both metrics are not ok (if we log it or compare it with other people), because they don't take into account the relation between `k` and `n`. If the model is ideal, `precision(k)` will decrease with increasing `k` (when `k > n`) and `recall(k)` is low with low `k` (when `k < n`). So logging or comparing those metrics with other people is meaningless.

[Wiki says](https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Precision_at_k), that we can limit the denominator by the available amount of `n` for `k`, replacing the denominator to `min(k, n_gt)`. So in terms of OML, `precision@k = cumsum(gt)(k) / min(k, n_gt)`. And if we define `precision@k` in the repo in such a way, the statement that `map@k = mean(precision@1 + precision@2 + ... + precision@k)` is not correct.

Formula with `rel(i)` factor is correct, but with `precision(k) = cumsum(gt)(k) / k`.

Reply:
I see. So, as of `precision@k` you take into account that `n` could be less than `k`, but as of `map@k` you don't take it into account. It looks a little odd though.

Reply:
@dapladoc last notes about `map`

Probably, we should name our modified precision with another name. But in order not to create new names, we decided to refer to the wiki with `precision@k`.

Reply:
Resolve it one more time, if it's ok

Reply:
Now it's completely ok. I understand the changes in the `precision` you've made and why you haven't made the same changes to the `map`, and this sounds very reasonable. The only problem is with the namings. I think this difference between your version of precision and the common one should be somehow describe in the docstrings.

I was using `MRR@k` and `Precision@k` (I used the version just like yours) in my project to track the position of the first relevant element among top `k` elements, and the precision.

Reply:
@dapladoc > I think this difference between your version of precision and the common one should be somehow describe in the docstrings.

Could you do this, please?